sys/can/device.c

```diff
@@ -0,0 +1,598 @@
+
+    for (int i = 0; i <= 1; i++) {
+        tseg2 = tseg + CAN_SYNC_SEG - (spt_nominal * (tseg + CAN_SYNC_SEG)) / 1000 - i;
+        tseg2 = clamp(tseg2, btc->tseg2_min, btc->tseg2_max);
+        tseg1 = tseg - tseg2;
+        if (tseg1 > btc->tseg1_max) {
+            tseg1 = btc->tseg1_max;
+            tseg2 = tseg - tseg1;
+        }
+
+        spt = 1000 * (tseg1 + CAN_SYNC_SEG) / (tseg + CAN_SYNC_SEG);
+        spt_error = max(spt, spt_nominal) - min(spt, spt_nominal);
+        if (spt <= spt_nominal && spt_error < min_spt_error) {
+            best_spt = spt;
+            min_spt_error = spt_error;
+            *p_tseg1 = tseg1;
+            *p_tseg2 = tseg2;
+        }
+
+        if (p_spt_error) {
+            *p_spt_error = min_spt_error;
+        }
+        DEBUG("tseg1=%" PRIu32 ", tseg2=%" PRIu32 ", spt_error=%" PRIu32 "\n",
+              tseg1, tseg2, spt_error);
+    }
+
+    return best_spt;
+}
+
+/*
+ *                   Nominal bit time (nbt) composed of 8 time quantum (tq)
+ * |<------------------------------------------------------------------------------------->|
+ * |                                                                                       |
+ * +----------+----------+-------------------------------------------+---------------------+
+ * | SYNC_SEG | PROP_SEG |                PHASE_SEG_1                |     PHASE_SEG_2     |
+ * +----------+----------+-------------------------------------------+---------------------+
+ * |                                                                 ^                     |
+ * |                                                    Sample point | at 75%              |
+ * |----------|----------|----------|----------|----------|----------|----------|----------|
+ * |   Time quanta                                                 6 | 2                   |
+ *
+ * Synchronization segment = always 1 tq
+ *                SYNC_SEG + PROP_SEG + PHASE_SEG1
+ * Sample point = --------------------------------
+ *                             nbt
+ *
+ * tseg1 = PROP_SEG + PHASE_SEG_1
+ * tseg2 = PHASE_SEG_2
+ * tseg = tseg1 + tseg2
+ * nbt = tseg + SYNC_SEG
+ *
+ */
+int can_device_calc_bittiming(uint32_t clock, const struct can_bittiming_const *timing_const,
+                              struct can_bittiming *timing)
+{
+    uint32_t spt; /* nominal sample point */
+    uint32_t spt_error;
+    uint32_t min_spt_error = UINT32_MAX;
+    uint32_t brp = 0;
+    uint32_t best_brp = 0;
+    uint32_t nbt;
+    uint32_t tseg;
+    uint32_t tseg1;
+    uint32_t tseg2;
+    uint32_t best_tseg = 0;
+    uint32_t rate;        /* current bitrate */
+    uint32_t rate_error;
+    uint32_t min_rate_error;
+
+    assert(timing != NULL && timing->bitrate != 0);
+    assert(timing_const != NULL);
+
+    /* Use recommended sample points */
+    if (timing->sample_point) {
+        spt = timing->sample_point;
+    }
+    else {
+        if (timing->bitrate > 800000) {
+            spt = 750;
```

Comment:
These are the sample point value recommended by the CiA (CiA 301). But I just discovered this was a bit outdated, last document recommends 87.5% for every baudrate up to 1Mbit/s.

I will fix that and add some comments in the code to explain.